Configure some VF flags if configured in net-attach-def

vpickard opened the pull request:
This PR is based off of PR#88, which pulled in an updated netlink lib that has functions to set the VF flags mentioned in this PR.

Question for folks, should we also cache existing values for these flags (spookchk, trust, etc) and restore them on cmdDel, similiar to what was done for MAC in PR#88?

Another similiar question, should there be code in ResetVFConfig() to also reset these above flags to default values?

vpickard commented:
The build for this PR is failing on "make test". I see this locally also, looks like I probably need to add some tests in sriov_test.go, right?

przemeklal commented on pkg/sriov/sriov.go:
```diff
@@ -309,7 +339,48 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
-	// 3. Set link rate
+	// 3. Set link rate.
+	if conf.MaxTxRate != nil {
+		if err = s.nLink.LinkSetVfTxRate(pfLink, conf.VFID, *conf.MaxTxRate); err != nil {
```
FYI - this option is deprecated, from the ip tools manual:
```
rate TXRATE -- change the allowed transmit bandwidth, in Mbps, for the specified VF.  Setting this parameter to
0 disables rate limiting.  vf parameter must be specified.  Please use new API max_tx_rate option instead.
```

vpickard replied:
Thanks, I saw this also. When looking at the netlink code for LinkSetVfTxRate(), the comment says that this is for the Max TX Bandwidth Alocation:

https://github.com/vishvananda/netlink/blob/2e4a68ee6c165085c6d708748b78cb8ac3c15ac9/nl/link_linux.go#L213

I also took a look at this during my testing. Here is the output from "ip link" when I had set the max_tx_rate to 50 Mbps:


5: em2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether ec:f4:bb:c0:bc:82 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 02:02:02:02:04:04, vlan 20, tx rate 50 (Mbps), max_tx_rate 50Mbps, spoof checking off, link-state auto, trust on, query_rss off
    vf 1 MAC 02:02:02:02:02:02, vlan 20, tx rate 50 (Mbps), max_tx_rate 50Mbps, spoof checking off, link-state auto, trust on, query_rss off
    vf 2 MAC 16:ba:a6:98:c0:fd, spoof checking on, link-state auto, trust off, query_rss off
    vf 3 MAC fe:4a:ec:08:51:ab, spoof checking on, link-state auto, trust off, query_rss off
    vf 4 MAC 26:a9:2b:2b:db:c1, spoof checking on, link-state auto, trust off, query_rss off
    vf 5 MAC 6e:e6:0d:fa:b1:2e, spoof checking on, link-state auto, trust off, query_rss off
    vf 6 MAC 4a:45:32:9e:9e:fc, spoof checking on, link-state auto, trust off, query_rss off


So, it appears this netlink call does set the max_tx_rate as specified. What is interesting, is that the "rate" attribute is also set. So, I think this is working as expected. Agree? Comments?

vpickard commented:
Thanks to @zshi-redhat for providing the mock test code for the new netlink calls. For reference, here is the link to his patch that I incorporated into this PR:
https://github.com/zshi-redhat/sriov-cni/commit/9589564b75e817da0a44fac9c13015a06da3570a

dankenigsberg commented:
Maybe I missed it: does this pr include documentation for the new config flags it exposes?

vpickard commented:
> Maybe I missed it: does this pr include documentation for the new config flags it exposes?

Hi,
No, I haven't included documentation for these config flags. 

These flags and doc can be found from "man ip link", in the VF section.
I can certainly add this information if you think that would be helpful. If that is the case, where would be the best place to add it in this repo?

Thanks,
Vic

dankenigsberg commented:
Maybe I'm wrong, but it seems that you don't have a blind pass-through of ip-link flags. If Linux kernel adds a new flag next year, you'd need to add it here.

Hence I think we should find a place to list the flags that you are adding with their expected capitalization.

moshe010 commented on pkg/sriov/sriov.go:
```diff
@@ -18,10 +19,17 @@ import (
 	LinkSetDown(netlink.Link) error
 	LinkSetNsFd(netlink.Link, int) error
 	LinkSetName(netlink.Link, string) error
+	LinkSetVfTxRate(netlink.Link, int, int) error
```
Vf support also min rate see [1]. It just not implemented in the netlink golang 

[1] - https://bugzilla.redhat.com/show_bug.cgi?id=1340914

vpickard replied:
Yes, I had added that as a comment in types.go. There are currently 3 flags that are not supported in netlink lib, min_tx_rate, query_rss, state.

vpickard commented:
> Maybe I'm wrong, but it seems that you don't have a blind pass-through of ip-link flags. If Linux kernel adds a new flag next year, you'd need to add it here.
> 
> Hence I think we should find a place to list the flags that you are adding with their expected capitalization.

OK, sounds reasonable. Are you thinking there should be some doc directory listing out flags that are supported and expected format? I could also add an example CRD to the images dir, or, perhaps better, create a deployment directory and add an example CRD there?

moshe010 commented on pkg/sriov/sriov.go:
```diff
@@ -245,8 +321,66 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
+	}
 
-	// 3. Set link rate
+	// 4. Set spoofchk flag
```
it seem you can do it in a loop gove ove all the attributes in conf and call the right netlink command

moshe010 commented on pkg/sriov/sriov.go:
```diff
@@ -18,10 +19,17 @@ import (
+	LinkSetVfSpoofchk(netlink.Link, int, bool) error
+	LinkSetVfTrust(netlink.Link, int, bool) error
+	LinkSetVfNodeGUID(netlink.Link, int, net.HardwareAddr) error
+	LinkSetVfPortGUID(netlink.Link, int, net.HardwareAddr) error
```
Do we need the LinkSetVfTxRate, LinkSetVfNodeGUID, LinkSetVfPortGUID? What is the point of this PRto expose all possible setting all the vf? What is the use-case for this PR?

vpickard replied:
Hi, 
Good questions, thank you!

The use case for this came about when testing DPDK/SRIOV with the testpmd app. By default, testpmd uses hardcoded MAC addresses. As such, the testpmd app was not able to pass packets, as the hardcoded MAC addresses didn't match the VF MAC addresses. So, I had to disable spoofchk for the packets to pass. 

While I was adding support for the spoofchk flag, I thought it would be good to add additional flags that are currently supported in the netlink lib. So, yes, the point of this PR would be to expose all possible, currently supported (netlink and device) VF flags. In the current state, this PR only adds support for a subset of the available VF flags (from "man ip link").

Now, DPDK based apps, including testpmd, have the command options, such as whitelist, to specify the MAC(s) to use, which would not require disabling spoofchk. 

But, the thinking is that there may be use cases where a DPDK app may desire to use a MAC other than the assigned MAC, which would require spoofchk to be disabled. 

Hope this is helpful, thanks for the input and comments.
Vic

moshe010 replied:
I understand the use-case for spoofcheck and trust mode, but I still wonder if we need to expose all possible VF flags or just add them when we have real use-case.

vpickard replied:
Sure, that is an option also. I can adjust this PR to only set those flags. I'm not aware of real-world use cases for the other flags, so I'm OK with that approach as well.

Any thoughts from others regarding just supporting spoofcheck and trust mode VF flags?

vpickard replied:
Although, I think the max_tx_rate may be useful, for rate limiting

moshe010 replied:
Yes I agree regarding the max_tx_rate we also added PR to https://github.com/vishvananda/netlink/pull/469 to add min rate so once it merged you can add that too.

vpickard replied:
Ok, sounds good. I will add the min_rate once that PR is merged

ahalim-intel commented:
@vpickard Thanks for the PR.
Thanks everyone for reviewing it. As pointed out by @moshe010, lets just add only configurations that are needed and not enabling everything without any strong use-case. Many of these configurations can be added later as well when they are needed.

Also, needs update the configuration section in the README.md as well.

dankenigsberg commented:
> > Maybe I'm wrong, but it seems that you don't have a blind pass-through of ip-link flags. If Linux kernel adds a new flag next year, you'd need to add it here.
> > Hence I think we should find a place to list the flags that you are adding with their expected capitalization.
> 
> OK, sounds reasonable. Are you thinking there should be some doc directory listing out flags that are supported and expected format? I could also add an example CRD to the images dir, or, perhaps better, create a deployment directory and add an example CRD there?

@vpickard I'd be content if you add your new flags to https://github.com/intel/sriov-cni#configuration-reference.

vpickard commented:
> > > Maybe I'm wrong, but it seems that you don't have a blind pass-through of ip-link flags. If Linux kernel adds a new flag next year, you'd need to add it here.
> > > Hence I think we should find a place to list the flags that you are adding with their expected capitalization.
> > 
> > 
> > OK, sounds reasonable. Are you thinking there should be some doc directory listing out flags that are supported and expected format? I could also add an example CRD to the images dir, or, perhaps better, create a deployment directory and add an example CRD there?
> 
> @vpickard I'd be content if you add your new flags to https://github.com/intel/sriov-cni#configuration-reference.

Thanks, I have updated the the README to include description of the flags, and sample usage.

vpickard commented:
> @vpickard Thanks for the PR.
> Thanks everyone for reviewing it. As pointed out by @moshe010, lets just add only configurations that are needed and not enabling everything without any strong use-case. Many of these configurations can be added later as well when they are needed.
> 
> Also, needs update the configuration section in the README.md as well.

Yes, thanks everyone for the review and comments, much appreciated.

I have updated the PR to just set the spoofchk, trust, and max_tx_rate flags. Doc/usage has been updated as well. 

Additionally, I rebased this PR to master (instead of PR#88). As a result, I had to update the glide config for the netlink library (I set it to the latest version), and that required an updated version of the unix pkg in golang.org. Here also, I set it to the latest version. Please let me know if there is some preferred way of selecting specific versions of these packages, other than choosing the latest and greatest.

Thanks,
Vic

vpickard commented:
The build failed from what appears to be some type of infra issue? I don't think I have access to trigger a rebuild/recheck......could someone with access please kick off a recheck/rebuild?

Thanks,
Vic


0.02s$ make lint
setting GOPATH...
updating dependencies...
/bin/sh: 1: glide: not found
Makefile:138: recipe for target 'glide.lock' failed
make: *** [glide.lock] Error 127
The command "make lint" exited with 2.

zshi-redhat commented:
> The build failed from what appears to be some type of infra issue? I don't think I have access to trigger a rebuild/recheck......could someone with access please kick off a recheck/rebuild?
> 
> Thanks,
> Vic
> 
> 0.02s$ make lint
> setting GOPATH...
> updating dependencies...
> /bin/sh: 1: glide: not found
> Makefile:138: recipe for target 'glide.lock' failed
> make: *** [glide.lock] Error 127
> The command "make lint" exited with 2.

I re-trigger the build once, it failed with the same error
I think the failure is due to dependencies in Makefile for each of lint/build/test etc to update vendor pkg.
The dependencies are not necessary as we will guarantee all the vendor pkg be updated to pass build.
I submit a [PR](https://github.com/intel/sriov-cni/pull/93) to remove the dependencies in Makefile (we did this also in sriov-network-device-plugin).

zshi-redhat commented:
@vpickard I noticed that glide.lock is updated but there is no vendor pkg updated in current PR. I assume there will be a follow-up commit for the vendor pkg update?

vpickard commented:
> @vpickard I noticed that glide.lock is updated but there is no vendor pkg updated in current PR. I assume there will be a follow-up commit for the vendor pkg update?

Thanks @zshi-redhat . I just pushed a separate commit to update the vendor packaging.

zshi-redhat commented on pkg/sriov/sriov.go:
```diff
@@ -247,6 +265,33 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
+	}
+
+	// 4. Set spoofchk flag
+	if conf.SpoofChk != "" {
```
Shall we reset VF config (`spoof check` and `trusted vf`) to original values or default values when VF is released?

vpickard replied:
I had asked the same questions when I first opened this PR, was hoping to get some input from folks.

I think, it would be good to reset to default values when the VF is released. So, in ResetVFConfig(), I could add code to set:

spoofchk -> on
trust -> off
max_tx_rate -> 0, which means no rate limiting.

I don't think it would be required to cache the original value and restore when VF is released, but if others have strong opinion that this is preferred, I could add that as well.

zshi-redhat replied:
I agree, we don't have to cache and restore, resetting to default value shall be fine. 
Just like what we did to vlan, it will be reset to 0.
And for security reason, we should reset spoof check to on and trusted vf to off .

vpickard replied:
OK, let me make that change, test it, and will push an updated patch with those changes.

vpickard replied:
Added another commit that sets the above flags to default values when VF is released.

zshi-redhat commented:
Tested the PR with i40evf on both kernel and DPDK modes, `spoofchk` and `trust` can be set and restored correctly.

vpickard commented:
Hi all,
Is there anything else that needs to be done before this PR can be merged? 

Thanks,
Vic

moshe010 commented:
@vpickard, Didi you tested in Mellanox cards?
We plan to test the functionality this week or start of next week ...

vpickard commented:
> @vpickard, Did you tested in Mellanox cards?
> We plan to test the functionality this week or start of next week ...

@moshe010, No, I have not yet, good to hear you can test this out on Mellanox.  Let's wait and make sure that goes well. If I get a chance to test on Mellanox, I will also update here.

Thanks

moshe010 commented:
tested on Mellanox card and it works as expected
